samples/custom_dispatch/vulkan/shaders/example_inline.mlir

```diff
@@ -52,13 +52,15 @@ module @example attributes {hal.device.targets = [#vulkan_target]} {
     %dim_i32 = arith.index_cast %dim : index to i32
 
     // Dispatch a basic `ret = lhs * rhs` shader.
-    %0 = hal.dispatch.extern "main"[%dim](%dim_i32, %arg0, %arg1) : (i32, tensor<?xf32>{%dim}, tensor<?xf32>{%dim}) -> tensor<?xf32>{%dim}
+    // Note that not all backends use names or the names are derived from
+    // ordinals so we include that (`:ordinal`).
+    %0 = hal.dispatch.extern "main":0[%dim](%dim_i32, %arg0, %arg1) : (i32, tensor<?xf32>{%dim}, tensor<?xf32>{%dim}) -> tensor<?xf32>{%dim}
+      // This host function is used to compute the XYZ workgroup count
+      // dispatched at runtime. It can query the %device for capabilities
+      // and limits (shared memory size, etc). The other arguments are the
+      // values passed in the dispatch operation (usually things like root
+      // output op tensor dimensions and other abstract values).
       count(%device: !hal.device, %workload: index) -> (index, index, index) {
```

Comment:
Same count region yes but not the same counts - the count region can be arbitrarily complex (it can query device features, have switches/conditions/etc). The condition regions below are to control which objects are used and that's N:M with the counts and we don't want to enforce 1:1.

Reply:
ah right, good point. I would expect cases though where the switch in the count region will be very similar to the object switch, in which case having good canonicalizations/cse will be important, but this makes sense, thanks for clarifying.